File: fc-chat/demo/part3/ch3/03-1-message-system-add-channel/src/main/java/net/prostars/messagesystem/entity/BaseEntity.java
package net.prostars.messagesystem.entity;

import jakarta.persistence.Column;
import jakarta.persistence.MappedSuperclass;
import jakarta.persistence.PrePersist;
import jakarta.persistence.PreUpdate;
import java.time.LocalDateTime;

@MappedSuperclass
public abstract class BaseEntity {
  
  @Column(name = "created_at", updatable = false, nullable = false)
  private LocalDateTime createAt;

  @Column(name = "updated_at", nullable = false)
  private LocalDateTime updatedAt;

  public LocalDateTime getCreateAt() {
    return createAt;
  }

  public LocalDateTime getUpdatedAt() {
    return updatedAt;
  }

  @PrePersist
  public void prePersist() {
    this.createAt = LocalDateTime.now();
    this.updatedAt = this.createAt;
  }

  @PreUpdate
  public void preUpdate() {
    this.updatedAt = LocalDateTime.now();
  }
}

File: fc-chat/demo/part3/with-push/src/main/kotlin/com/yscorp/withpush/messagesystem/entity/BaseEntity.kt
package com.yscorp.withpush.messagesystem.entity

import jakarta.persistence.Column
import jakarta.persistence.MappedSuperclass
import jakarta.persistence.PrePersist
import jakarta.persistence.PreUpdate
import java.time.LocalDateTime

@MappedSuperclass
abstract class BaseEntity {
    @Column(name = "created_at", updatable = false, nullable = false)
    var createAt: LocalDateTime? = null
        private set

    @Column(name = "updated_at", nullable = false)
    var updatedAt: LocalDateTime? = null
        private set

    @PrePersist
    fun prePersist() {
        this.createAt = LocalDateTime.now()
        this.updatedAt = this.createAt
    }

    @PreUpdate
    fun preUpdate() {
        this.updatedAt = LocalDateTime.now()
    }
}
